File: Python/1-range.py
class Range:
	
	def __init__(self, first, second=None, step=1):
		if second is None:
			self.start = 0
			self.current = 0
			self.end = first
			self.step = step
		else:
			self.start = first
			self.current = first
			self.end = second
			self.step = step

	def __iter__(self):
		return self

	def __next__(self):
		result = self.current
		if (result >= self.end and self.step > 0
		 or result <= self.end and self.step < 0):
			raise StopIteration
		else:
			self.current += self.step
			return result

def print_space(str):
	print("{}".format(str), end=' ')

for i in Range(10):
	print_space(i)
print()
for i in Range(3, 18):
	print_space(i)
print()
for i in Range(2, 15, 2):
	print_space(i)
print()
for i in Range(10, 0, -1):
	print_space(i)

File: JavaScript/1-range.js
'use strict';

class Range {
  constructor(first, second, step = 1) {
    if (second === undefined) {
      this.start = 0;
      this.end = first;
    } else {
      this.start = first;
      this.end = second;
    }
    this.step = step;
  }

  [Symbol.iterator]() {
    const { start, end, step } = this;
    let current = start;
    return {
      next() {
        if (current === undefined || step === 0) {
          return { value: undefined, done: true };
        }

        if (
          current >= end && step > 0 ||
          current <= end && step < 0
        ) {
          const value = current;
          current = undefined;
          return { value, done: true };
        }

        const value = current;
        current += step;
        return { value, done: false };
      }
    };
  }
}

console.log([
  [...new Range(10)],
  [...new Range(3, 18)],
  [...new Range(2, 15, 2)],
  [...new Range(10, 0, -1)],
]);
